### PR TITLE
LargeTypesReg2Mem: Add a new heuristic that trys harder to keep large     values on the stack

### PR DESCRIPTION
**Pull request by aschwaighofer**
This heuristic can be enabled by passing -Xfrontend  -enable-aggressive-reg2mem.

rdar://123916109

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci test

**Comment by aschwaighofer** [comment]
@swift-ci test macOS

**Comment by aschwaighofer** [comment]
@swift-ci smoke test macOS

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/DeadStoreElimination.swift [hunk 1]
```diff
@@ -75,7 +75,7 @@ let deadStoreElimination = FunctionPass(name: "dead-store-elimination") {
 }
 
 private func tryEliminate(store: StoreInst, complexityBudget: inout Int, _ context: FunctionPassContext) {
-  if !store.hasValidOwnershipForDeadStoreElimination {
+  if !store.hasValidOwnershipForDeadStoreElimination || !store.source.type.shouldExpand(context) {
```
Comment (eeckstein) [inline SwiftCompilerSources/Sources/Optimizer/FunctionPasses/DeadStoreElimination.swift:78]: Can you add a short comment why preventing dead-store elimination of not-shouldExpand-types is beneficial?

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/RedundantLoadElimination.swift [hunk 2]
```diff
@@ -97,6 +97,9 @@ private func eliminateRedundantLoads(in function: Function, ignoreArrays: Bool,
         {
           continue
         }
+        if !load.type.shouldExpand(context) {
```
Comment (eeckstein) [inline SwiftCompilerSources/Sources/Optimizer/FunctionPasses/RedundantLoadElimination.swift:100]: The same here

File: SwiftCompilerSources/Sources/Optimizer/Utilities/OptUtils.swift [hunk 3]
```diff
@@ -809,3 +809,12 @@ extension CheckedCastAddrBranchInst {
     }
   }
 }
+
+extension Type {
+  func shouldExpand(_ context: some Context) -> Bool {
```
Comment (eeckstein) [inline SwiftCompilerSources/Sources/Optimizer/Utilities/OptUtils.swift:814]: Can you add a comment what `shouldExpand` actually means? E.g. copy the comment from the C++ implementation

**Review by eeckstein — COMMENTED** [review]
Nice!

@aschwaighofer can you add some comments as I suggested? Otherwise it's really hard to figure out why this checks are done.

**Pull request merged**